File: source/python/69. Sqrt(x).py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 0:
            return 0
        if x == 1:
            return 1
        left = 1
        right = x
        while left <= right:
            mid = (left + right) // 2
            square = mid * mid
            if square == x:
                return mid
            elif square < x:
                left = mid + 1
            else:
                right = mid - 1
        return right


def test():
    s = Solution()
    assert s.mySqrt(4) == 2
    assert s.mySqrt(8) == 2

    print("All tests passed!")


if __name__ == "__main__":
    test()

File: source/cpp/69. Sqrt(x).cpp
class Solution {
public:
    int mySqrt(int x) {
        if (x == 0) return 0;
        if (x == 1) return 1;

        auto left = 1ULL;                           // 1UUL is used to avoid overflow
        auto right = x;
        while (left <= right) {
            auto mid = (left + right) / 2UL;
            auto square = (long long)mid * mid;
            if (square == x) return mid;
            if (square < x)
                left = mid + 1;
            else
                right = mid - 1;
        }

        return right;
    }
};